### DISCUSSION TITLE
Fastlane / Match + iOS + Azure DevOps

### DISCUSSION THREAD

**Discussion by CalvinAllen** [fastlane/fastlane#16831] [Q&A]
Hey folks,

Hoping somebody can help me out here.  I've got about 8 hours into trying to get fastlane and match working in our AzureDevOps pipelines for our iOS build, and I am having zero luck.

I've tried so many things at this point, such as: `setup_ci`, performing similar actions like `setup_ci` directly in the fastfile, toggling various options on/off (readonly, refresh, etc), setting the various ENV variables for match and keychain), going from automatic signing to manual (and back again), running various bash scripts, etc.  I always end up with the same error (paraphrased):

> No code signing identities could be found

What baffles me, even further, is that there are AzureDevOps tasks for installing a certificate and provisioning profile, so it *has* to work in that enviornment, right?  Those tasks require having the certificate and profile uploaded into the library as secure files, which doesn't work well with `match` (all my stuff syncs to S3).

Does anybody out there have this working, or have any ideas?

Thanks!

**Comment by joshdholtz** [comment]
@CalvinAllen 👋 Hey! Would you be able to post part of what your `Fastfile` looks like and your console output from Azure? That would be a great place to help us start looking into what's going on 😊

**Reply by CalvinAllen** [reply]
Yes, I can, although it might be delayed a bit; I'm out of town for a family emergency at the moment.  

Also, I've tried so many different things, I'll probably start this with "the basics", and provide the output along the way as we try different things.

**Reply by joshdholtz** [reply]
I hope everything will be okay! You do you, first. I'd be happy to help whenever its convenient for you 👍

**Comment by CalvinAllen** [comment]
Okay, finally got some time to retrigger everything.

Heres the error I'm getting:

```
[00:28:31]: ▸ ❌  error: There are no accounts registered with Xcode. Add your developer account to Xcode (in target '<app name>' from project '<app name>')
[00:28:31]: ▸ ❌  error: No profiles for 'com.<company>.<app name>.dev' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.<company>.<app name>.dev'. (in target '<app name>' from project '<app name>')
```

These are the two pertinent steps of the pipeline:

```
- task: Bash@3
  displayName: 'Fastlane Match ...'
  inputs:
    targetType: 'inline'
    script: |
      fastlane ios certificates

- task: Bash@3
  displayName: 'Build & Deploy iOS Development to AppCenter...'
  inputs:
    targetType: 'inline'
    script: |
      npm run ios-deploy:dev
```

The `ios certificates` lane is defined as:

```
  desc "Synchronize certificates and profiles"
  lane :certificates do |options|
     setup_ci

     match(app_identifier: 'com.<company>.<app name>.dev', type: 'development', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>.dev', type: 'adhoc', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>.dev', type: 'appstore', force_for_new_devices: true, readonly: true)

     match(app_identifier: 'com.<company>.<app name>.qa', type: 'development', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>.qa', type: 'adhoc', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>.qa', type: 'appstore', force_for_new_devices: true, readonly: true)

     match(app_identifier: 'com.<company>.<app name>', type: 'development', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>', type: 'adhoc', force_for_new_devices: true, readonly: true)
     match(app_identifier: 'com.<company>.<app name>', type: 'appstore', force_for_new_devices: true, readonly: true)
  end
```

The pertinent piece of `build` is:

```    
    @method = (@scheme == "Development") ? "development" : "ad-hoc"
    @method = (@scheme == "QA") ? "ad-hoc" : @method
    @method = (@scheme == "Production") ? "app-store" : @method

    if(options[:methodOverride])
        @method = options[:methodOverride]
    end

    gym(
      scheme: @scheme,
      workspace: "./ios/<app name>.xcworkspace",
      clean: true,
      silent: true,
      xcargs: "-allowProvisioningUpdates",
      export_method: @method
  )
```

The `setup_ci` reported:

```
[00:25:48]: ----------------------
[00:25:48]: --- Step: setup_ci ---
[00:25:48]: ----------------------
[00:25:48]: Creating temporary keychain: "fastlane_tmp_keychain".
[00:25:48]: $ security list-keychains -d user
[00:25:48]: ▸ "/Users/runner/Library/Keychains/fastlane_tmp_keychain-db"
[00:25:48]: Found keychain '/Users/runner/Library/Keychains/fastlane_tmp_keychain-db' in list-keychains, adding to search list skipped
[00:25:48]: Enabling match readonly mode.
```
And this is the output at the very end of the rest of the certificates lane:

```
+------+----------+-------------+
|       fastlane summary        |
+------+----------+-------------+
| Step | Action   | Time (in s) |
+------+----------+-------------+
| 1    | setup_ci | 0           |
| 2    | match    | 2           |
| 3    | match    | 2           |
| 4    | match    | 1           |
| 5    | match    | 2           |
| 6    | match    | 1           |
| 7    | match    | 1           |
| 8    | match    | 2           |
| 9    | match    | 1           |
| 10   | match    | 1           |
+------+----------+-------------+

[00:26:08]: fastlane.tools finished successfully 🎉
```

Although that says successful, there was an error logged at the start:

```
00:25:51]: 🔓  Successfully decrypted certificates repo
[00:25:51]: Installing certificate...
security: SecKeychainSearchCopyNext: The specified item could not be found in the keychain.
[00:25:52]: There are no local code signing identities found.
You can run `security find-identity -v -p codesigning fastlane_tmp_keychain` to get this output.
This Stack Overflow thread has more information: https://stackoverflow.com/q/35390072/774.
(Check in Keychain Access for an expired WWDR certificate: https://stackoverflow.com/a/35409835/774 has more info.)
[00:25:52]: Setting key partition list... (this can take a minute if there are a lot of keys installed)
[00:25:52]: security: SecItemCopyMatching: The specified item could not be found in the keychain.
[00:25:52]: Setting key partition list... (this can take a minute if there are a lot of keys installed)
```

I'm using `s3` as the storage mechanism and everything works locally.  In Azure, I'm using the `macos-latest` image (which is, right now, the same as `macos-10-15`, and [has all the software in it as mentioned here](https://github.com/actions/virtual-environments/blob/master/images/macos/macos-10.15-Readme.md)

Okay, I think that covers everything :D

**Reply by joshdholtz** [reply]
Thanks for posting this!

## Looks like you need to disable automatic code signing
`error: There are no accounts registered with Xcode.` shows me that your Xcode project has automatic code signing enabled. This will fight with _match_ which does manual code signing.

❓ Can you verify that you have your build settings set to manual signing

## If/when you have manual signing, make sure to have build signing set properly
1. `PROVISION_PROFILE_SPECIFIER` should be set the name of the profile that _match_ has made for you
2. `CODE_SIGN_IDENTITY` needs to be set to either `Apple Distribution`, `Apple Development`, `iPhone Distribution`, `iPhone Development` (depends on what type of certificate is used for your provisioning profiles)
3. `CODE_SIGN_STYLE` needs to be set to `Manual`

**Reply by CalvinAllen** [reply]
When I change to manual signing, I start getting errors (locally) related to the provisioning profile selection:

```
[10:41:13]: ▸ ❌  error: "<app name>" requires a provisioning profile with the Push Notifications feature. Select a provisioning profile in the Signing & Capabilities editor. (in target '<app name>' from project '<app name>')
```

If I go in to xcode and set the profile to `match AdHoc com.<company>.<app name>.dev` then I start getting errors such as:

```
[10:49:43]: ▸ ❌  error: Signing certificate is invalid. Signing certificate "Apple Distribution: <apple account name> (<apple team ID>)", serial number "<serial number>", is not valid for code signing. It may have been revoked or expired. (in target '<app name>' from project '<app name>')
```

I'm not sure whats going on 😕, all the certificates are valid (either expires 1 year from yesterday, or this coming December).

When you say, "If/when you have manual signing, make sure to have build signing set properly" - is that a manual step I need to do, or do those get set appropriately during `gym` and manual matching?

Thanks!

EDIT:

Have I conflated things by trying to match Development/Adhoc/AppStore on all bundle ids?  I have 2 certificates, 3 bundle IDs, and 9 profiles.

**Reply by CalvinAllen** [reply]
Actually, I may have answered my own questions.  Sorry, I inherited a lot of this stuff, and mobile dev wasn't my focus prior to that.

**Reply by CalvinAllen** [reply]
Okay, @joshdholtz, that solved my problem, Azure just completed the first build and deployed it to AppCenter.

I am still a bit confused, though, on how to build the other variations.

For instance, when I build the "dev", `com.<company>.<app name>.dev` and match that to the right ad-hoc profile (in Xcode), it all works fine.  When I run my "qa" lane, which builds `com.<company>.<app name>.qa` and requires a different profile, I can't get it to build (even locally), because I'm swapping the scheme, bundle identifier, etc.  Whats the right way to approach this?  I can't see how I can a different profile for a different bundle identifier.

**Reply by joshdholtz** [reply]
@CalvinAllen I’m glad you got it working! And also inheriting an Xcode project that has multiple environments/schemes and getting signing to working is no easy task so 🏆 for figuring this all out!

There are multiple approaches you can take solve this.

## Option 1 - Build out different configurations/xcconfig for each environment
Xcode projects come with a `Development` and `Release` scheme by default. This works for simple use cases but doesn’t work when you have multiple environment you need to configure. What you can do is add configurations for `Dev`, `Staging`, and `Production`. I usually do this approach with my projects. It allows you to specific different bundle identifiers, code signing identity, and provisioning profile specifier per environment.

There is a good tutorial on NSScreencast on how to do this:
- Part 1: https://nsscreencast.com/episodes/154-xcconfig-files
- Part 2: https://nsscreencast.com/episodes/155-xcconfig-files-part-2

An approach like this then allows you to change the `configuration` option in the `gym` / `build_ios_app` action to build for your different environments. You then just need to make sure you pass the correct bundle identifier to `match`.

## Option 2 - Change build settings on CI on the fly
If you don’t want to mess with Xcode configurations, you can also just change all the build settings on the fly with - http://docs.fastlane.tools/actions/update_code_signing_settings/#update_code_signing_settings

This is a little bit messier in terms of automation/CI but it is _pretty okay_ to change these on the fly on the CI 🤷‍♂️

**Reply by CalvinAllen** [reply]
@joshdholtz Ah, both of those make sense.  I saw the `update_code_signing_settings` action, but it wasn't clear to me (until now!) why I would have needed it. Thanks again for all your help, really appreciate it!

**Reply by eduardosystem** [reply]
I have an iOS and Android deployment mobile app. It would be great if they showed the processes to start the application from a repository using Fastlane. With graphics if possible and thus be able to find myself in the same case that is found to give a solution.

**Comment by emreond** [comment]
Hello. I'm also facing problems with using match with Azure Devops Pipelines. I'm getting timeout error when trying to clone the certificates repo.

**Reply by maxx0r** [reply]
If you haven't solved this yet; i have this every now and then and i have to re-add the ssh key

**Reply by emreond** [reply]
Hello. Can you please share the solution with me?

**Reply by mateszlanka** [reply]
Try to change the git url for `fastlane match` from `ssh` to `https`. Once you're using `https`, you will be able to set an Environment Variable in Azure Pipelines which will grant you access to you Match repository if the said repo is under the same organization as you are.
```
env:
  MATCH_GIT_BEARER_AUTHORIZATION: $(System.AccessToken)
```

**Reply by ycfreeman** [reply]
help i've banging my head for like 2 days for this and still haven't got it working. here's my match, fast files 

i've already set Pipeline permissions to Open
and grant build service user contribution  

Matchfile
```
git_url("https://miau2020@dev.azure.com/miau2020/uni-app/_git/certs")

storage_mode("git")
```

azure-pipelines.yml
```
...
  - script: |
      bundle install
      bundle exec fastlane ios certificates
    displayName: "iOS match"
    env:
      MATCH_GIT_BEARER_AUTHORIZATION: $(System.AccessToken)
...
```

Fastfile
```
  \desc "Fetch certificates and provisioning profiles"
  lane :certificates do
    match(app_identifier: app_identifier, type: "appstore", readonly: true)
  end
```

and im getting
```
fatal: repository 'https://dev.azure.com/miau2020/uni-app/_git/certs/' not found
```

**Reply by emreond** [reply]
Hey @ycfreeman . What I also done in my fastfile was; Try this :) 

``` 
      match(
      type: "adhoc",
      force_for_new_devices: true,
      git_basic_authorization: Base64.strict_encode64("myTeamName:#{ENV['AZURE_TOKEN']}"),
      git_url:  "https://#{ENV['AZURE_TOKEN']}@dev.azure.com/myTeamName/ProjectName/_git/ios_certificates", 
      readonly: true
     )
```

**Comment by httran13** [comment]
> Hey @ycfreeman . What I also done in my fastfile was; Try this :)
> 
> ```
>       match(
>       type: "adhoc",
>       force_for_new_devices: true,
>       git_basic_authorization: Base64.strict_encode64("myTeamName:#{ENV['AZURE_TOKEN']}"),
>       git_url:  "https://#{ENV['AZURE_TOKEN']}@dev.azure.com/myTeamName/ProjectName/_git/ios_certificates", 
>       readonly: true
>      )
> ```

Hey thanks, im still a bit confuse, can you share what AZURE_TOKEN is? from my understanding the format is https://<org>@dev.azure.com
